Decisions/scs-XXXX-v1-requirements-for-container-registry.md

```diff
@@ -0,0 +1,292 @@
+  - Ensure system uptime even in the event of a failure
+- Registry replication
+  - Replication allows users to replicate container images between registries of the same instances and between registries of different instances as well
+- Proxy cache (pull-through cache)
+  - Proxy cache allows you to use a container registry to proxy and cache images from a target public or private registry
+- Quota management
+  - Control over resource use
+- Garbage collection
+  - Removing blobs from the filesystem when they are no longer referenced by a manifest
+- Retention policy
+  - Reduce the number of image tags, many of which might not be required after a given time or once a subsequent image tag has superseded them
+
+**Desirable features**
+
+- Additionally supported artifacts
+  - Additional artifacts that the registry is able to store in addition to OCI artifacts, e.g. Java, Node.js, or Python packages
+- Integration possibilities
+  - Ability to cooperate with another software solution in order to improve own feature set (e.g. integration of P2P solution for improving container image distribution (download speed and stability, high scalability ...))
+
+Refer to the table of evaluated projects with their features. Note that only container
+registry implementations that passed the OSS health stage (Harbor, Quay, and Dragonfly)
+are evaluated here.
+
+| Features                       | Harbor                                        | Quay                                                                | Dragonfly                |
+|--------------------------------|-----------------------------------------------|---------------------------------------------------------------------|--------------------------|
+| Audit Logs                     | ✓                                             | ✓                                                                   | ✗                        |
+| Authentication                 | ✓ Local database, LDAP, OIDC, UAA             | ✓ Local database, LDAP, Keystone, JWT                               | ✓ Local database         |
+| Authorization                  | ✓                                             | ✓                                                                   | ✓                        |
+| Automation                     | ✓ Webhooks                                    | ✓ Webhooks, building images                                         | ✗                        |
+| Vulnerability scanning         | ✓ Trivy, Clair                                | ✓ Clair                                                             | ✗                        |
+| Content Trust and Validation   | ✓ Notary, Cosign                              | ✓ Cosign                                                            | ✗                        |
+| Multi-tenancy                  | ✓                                             | ✓                                                                   | ✓                        |
```

Comment:
I think the multi-tenancy bit may use a little bit more consideration.

For example, AFAIU, Harbor is indeed multi-tenancy-capable regarding RBAC controls and UI, making it a candidate for hosting a singleton "registry as a service" instance operated by a CSP.

However AFAIU, e. g. attributing used resources to each tenant for e. g. billing purposes is quite difficult, as Harbor operates on a single storage backend.

This specific aspect is e. g. addressed by https://github.com/sapcc/keppel#overview (a comparison between them and harbor is done directly in their README).

Not trying to say that this *necessarily* changes the overall picture. But there are multiple considerations regarding multi-tenancy and metering/billing is one of them; For the "registry as a service" use case, it may be one of the most important aspects from CSP-perspective and should be mentioned in this decision record (even if only as a design trade-off).

Reply:
Hello @joshmue

I got your points that may be difficult to charge e.g. storage in the case of Harbor multi-tenant as-a-service container registry solution (as Harbor operates on a single storage backend)

In my point of view, there are two options for how we can look at that:

1. CSPs may offer the whole Harbor instance as-a-service per user
   - In this case, the CSP may charge the whole storage to the user, because there is no shared one
   - Based on our research I would say that this is a common option across CSPs that offer container registry as-a-service based on the Harbor project, e.g. [container-registry.com](https://container-registry.com/#pricing), [OVH cloud](https://www.ovhcloud.com/en/public-cloud/managed-private-registry/). OVH even developed a [harbor-operator](https://blog.ovhcloud.com/managing-harbor-at-cloud-scale-the-story-behind-harbor-kubernetes-operator/) for that purpose, so they can control multiple harbor instances in a simple way
2. CSPs may offer multi-tenant Harbor instance as-a-service, hence one Harbor instance is shared across multiple users
   - Each user gets a separate tenant/project, e.g. Free plan for [container-registry.com](https://container-registry.com/#pricing)
   - The deciding aspect for metering/billing here might be quotas. Quota may be set per tenant/project, and e.g. the storage could be charged based on it. See also [Harbor project quotas](https://goharbor.io/docs/2.7.0/administration/configure-project-quotas/)

Overall I do not see the benefit here from considering various use cases in the context of multi-tenancy (as they are many and IMO it is out of the ADR scope). But I agree with you that would be precise to mention that the multi-tenancy in the projects like Harbor is not implemented on the storage level, but only on the virtual "project" level.

Reply:
Thanks for mentioning it!

To explain my reasoning behind stressing this aspect:
I guess this decision record should not only find the container registry software which is *generally* the best one out there, but also evaluate it from SCS perspective and prospective use cases.

As you wrote in the "Motivation" part, we have two/three main use cases:

> The specific use cases should be discussed, but overall the CSPs could offer a private container registry as a service or the CSPs could offer a recipe for customers to deploy the private registry themselves utilizing the CSP infrastructure.
> ...
> In both cases, the private container registry should be SCS-compliant

**Use case 1**: CSP's offering a managed "container registry as a service", using recipes maintained by SCS
**Use case 2**: Users managing a container registry themselves, using recipes maintained by SCS [^1]
**Use case 3**: Defining an "SCS compliant container registry" to make switching registries easy

The third use case [IMHO needs a bit more of clarification](https://github.com/SovereignCloudStack/standards/pull/212/files#r1107046124), but should have value for end users, as long as there are managed or self-hostable solutions ready to use, which are kind-of interchangeable.

Only considering the second use case, SCS IMHO can offer only little benefits over upstream recipes. There is no strict separation of personas/roles involved here, making the feature list the most important aspect. Hence, for this use case, it really only matters, which software is *generally* the best one.

Considering the first use case, SCS can offer a lot of value to both CSP's and end users by providing a ready-to-go "aaS" solution, while having some deployment automation (for use case 2) as by-product.
I'd consider that one the primary use case (together with use case 3).
From CSP perspective, the multi-tenancy/quota/... capabilities may outweigh any other feature, as different approaches on multi-tenancy most likely will have massive implications on pricing, product strategy and technical architecture.
As you mentioned, there are two general approaches to multi-tenancy:

> 1. CSPs may offer the whole Harbor instance as-a-service per user
>    - In this case, the CSP may charge the whole storage to the user, because there is no shared one
>    - Based on our research I would say that this is a common option across CSPs that offer container registry as-a-service based on the Harbor project, e.g. [container-registry.com](https://container-registry.com/#pricing), [OVH cloud](https://www.ovhcloud.com/en/public-cloud/managed-private-registry/). OVH even developed a [harbor-operator](https://blog.ovhcloud.com/managing-harbor-at-cloud-scale-the-story-behind-harbor-kubernetes-operator/) for that purpose, so they can control multiple harbor instances in a simple way
> 2. CSPs may offer multi-tenant Harbor instance as-a-service, hence one Harbor instance is shared across multiple users
>    - Each user gets a separate tenant/project, e.g. Free plan for [container-registry.com](https://container-registry.com/#pricing)
>    - The deciding aspect for metering/billing here might be quotas. Quota may be set per tenant/project, and e.g. the storage could be charged based on it. See also [Harbor project quotas](https://goharbor.io/docs/2.7.0/administration/configure-project-quotas/)

Both approaches differ a lot from each other.

In my head there is the following analogy: OpenStack.
Like e. g. Harbor, OpenStack can be deployed by an operator for each of their users separately. AFAIK, this usually is not done, as OpenStack is multi tenancy capable and sharing a single OpenStack installation allows to share resources, hence saving resources, hence saving money of the operator and the users.

As such (I imagine), when a CSP starts a cloud from scratch, and they evaluate different cloud solutions, they would most likely see lack of multi-tenancy not as a minor implementation detail, but as a potential deal breaker.
The same **could** be the case for container registries.

I'm not trying to say that e. g. Harbor's multi tenancy features are insufficient (in fact, quotas alone may already be enough even for billing), but I'm trying to say that this **may** be a very important factor for CSP's (AFAIK SCS's primary "consumer group") or even users who are not ready to pay for a relatively large amount of compute to store a few container images.

I hope some CSP's are able to contribute their views on this.

[^1]: Sidenote: Who should really maintain the recipes? In the "Motivation", it says that the CSP should do it, to enable the user to deploy the registry on the CSP itself. This would (maybe I'm over-interpreting here) not mandate such recipes to be provider independent, hence making migration difficult. Avoiding such lock-in is an SCS priority. Maybe this could be clarified.

Reply:
Thank you for the detailed explanation of your thoughts.  Overall, They make sense to us. 

> .. Who should really maintain the recipes? In the "Motivation", it says that the CSP should do it  ...

Agree, the recipes should be CSP-agnostic and maintained by SCS. I mentioned that in the fixup commit 67b80cdae61d0b8f945af24cf38db23ee610d9a5
 
> I'm not trying to say that e. g. Harbor's multi tenancy features are insufficient (in fact, quotas alone may already be enough even for billing), but I'm trying to say that multi-tenancy in general may be a very important factor for CSP's ...

Agree, the question could be here whether Harbor is able to operate as a single (potentially large) multi-tenant instance for an undefined number of users. 
- @chess-knight picked out an interesting Harbor [issue](https://github.com/goharbor/community/issues/115) that summarizes real-world Harbor deployments by various companies
- There is a [comment](https://github.com/goharbor/community/issues/115#issuecomment-786308471) that mentioned CSP with 5k clients (not sure whether all clients shared one Harbor instance). Would be great if @Vad1mo could bring some light into it

Reply:
> 1. CSPs may offer the whole Harbor instance as-a-service per user
>    
>    * In this case, the CSP may charge the whole storage to the user, because there is no shared one
>    * Based on our research I would say that this is a common option across CSPs that offer container registry as-a-service based on the Harbor project, e.g. [container-registry.com](https://container-registry.com/#pricing), [OVH cloud](https://www.ovhcloud.com/en/public-cloud/managed-private-registry/). OVH even developed a [harbor-operator](https://blog.ovhcloud.com/managing-harbor-at-cloud-scale-the-story-behind-harbor-kubernetes-operator/) for that purpose, so they can control multiple harbor instances in a simple way
> 2. CSPs may offer multi-tenant Harbor instance as-a-service, hence one Harbor instance is shared across multiple users
>    
>    * Each user gets a separate tenant/project, e.g. Free plan for [container-registry.com](https://container-registry.com/#pricing)
>    * The deciding aspect for metering/billing here might be quotas. Quota may be set per tenant/project, and e.g. the storage could be charged based on it. See also [Harbor project quotas](https://goharbor.io/docs/2.7.0/administration/configure-project-quotas/)
> 

We are currently running option 1 and 2. The main reason for having 1 & 2 is that 1 has a mandatory resource baseline on CPU/RAM/DB connections per tenant, so we need to price that in. Option 2 has a nice linear pricing on the other hand. 

Option 2 needs glue code and 3rd party systems to make it work, in a soft multi-tenant mode.

In the next ~6 months, we will only offer Option 1, as we now have fully multi-tenant capable instances, without allocating CPU/RAM per tenant.

Reply:
>  In the next ~6 months, we will only offer Option 1, as we now have fully multi-tenant capable instances, without allocating CPU/RAM per tenant.

Could you please explain this in detail? (Option 1 = whole Harbor instance (including storage layer) per tenant, i.e. single-tenant instance)

Reply:
> Could you please explain this in detail? (Option 1 = whole Harbor instance (including storage layer) per tenant, i.e. single-tenant instance)

As you guessed it: Whole Harbor instance (including storage layer) per tenant, but on a mutli-tenant instance. So, hence no additional CPU/RAM/DB connections per tenant.

Reply:
Is the multi-tenant instance you mentioned Kubernetes?

Reply:
no, multi-tenant instance (deployment) of Harbor.